beamer/middleware.py

```diff
@@ -41,3 +51,221 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+
+            if rate_limited:
+                # Even after _RATE_LIMIT_PERIOD, we are still being rate-limited by
+                # the RPC provider so there is nothing we can really do.
+                raise RuntimeError("rate limit period exceeded: %s" % rpc)
+            else:
+                assert response is not None
+                if time.time() > state.rate_limit_end:
+                    # Rate limit period ended so we now enter the tapering period,
+                    num_waiting_on_lock = state.num_waiting_on_lock.load()
+                    log.debug(
+                        "Exiting rate limiting mode",
+                        thread=threading.current_thread().name,
+                        rpc=rpc,
+                    )
+                    state.rate_limit_end = None
+                    state.taper_counter = 0
+                    state.taper_counter_max = num_waiting_on_lock
+
+                return response
+        else:
+            if state.taper_counter_max > 0:
+                # We are in tapering period.
+                request_delay = _RATE_LIMIT_REQUEST_DELAY / 2**state.taper_counter
+                state.taper_counter += 1
+                if state.taper_counter == state.taper_counter_max:
+                    # We have reached the end of tapering period.
+                    state.taper_counter = 0
+                    state.taper_counter_max = 0
+                    log.debug(
+                        "Exiting tapering mode", thread=threading.current_thread().name, rpc=rpc
+                    )
+
+                time.sleep(request_delay)
+
+            # If we get rate limited by the RPC, start a new rate limiting
+            # period by setting state.rate_limit_end and try sending the
+            # request again after _RATE_LIMIT_REQUEST_DELAY. Otherwise,
+            # simply return the response.
+            rate_limited, response = _try_make_request(make_request, method, params)
+            if rate_limited:
+                log.debug(
+                    "Entering rate limiting mode", thread=threading.current_thread().name, rpc=rpc
+                )
+                state.rate_limit_end = time.time() + _RATE_LIMIT_PERIOD
+                time.sleep(_RATE_LIMIT_REQUEST_DELAY)
+            else:
+                assert response is not None
+                return response
+
+
+def _rate_limiter(
+    method: RPCEndpoint,
+    params: Any,
+    make_request: _MakeRequest,
+    w3: Web3,
+    state: _RateLimiterState,
+) -> RPCResponse:
+    if hasattr(_RATE_LIMITER_TLD, "entered"):
+        # We already entered this function once and likely took the lock,
+        # so just make a request immediately to avoid deadlocks.
+        # The cause of a second entry into this function is most probably
+        # an eth_chainId call made as part of our call to make_request.
+        return make_request(method, params)
```

Comment:
just to clarify: We can call `make_request()` here because we are sure, that this is call within another call which comes from `_try_make_request()` and thus an exception will bubble up and be caught in `_try_make_request()`?

Reply:
Yes.